File: src/tools/registry.ts
import type { ChatCompletionTool } from "openai/resources/chat/completions";
import { zodToJsonSchema } from "../utils/zod-to-json-schema.js";
import type { Tool } from "./types.js";

const tools = new Map<string, Tool>();

export function registerTool(tool: Tool): void {
  tools.set(tool.name, tool);
}

export function getTool(name: string): Tool | undefined {
  return tools.get(name);
}

export function getAllTools(): Tool[] {
  return Array.from(tools.values());
}

export function toolToFunction(tool: Tool): ChatCompletionTool {
  return {
    type: "function",
    function: {
      name: tool.name,
      description: tool.description,
      parameters: zodToJsonSchema(tool.parameters),
    },
  };
}

File: src/tools/registry.py
from __future__ import annotations

from typing import Any

from src.tools.types import Tool

_tools: dict[str, Tool] = {}


def register_tool(tool: Tool) -> None:
    _tools[tool.name] = tool


def get_tool(name: str) -> Tool | None:
    return _tools.get(name)


def get_all_tools() -> list[Tool]:
    return list(_tools.values())


def tool_to_function(tool: Tool) -> dict[str, Any]:
    return {
        "type": "function",
        "function": {
            "name": tool.name,
            "description": tool.description,
            "parameters": tool.parameters,
        },
    }
